[FSDP][Easy] Reword device placement warning

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #77853 [FSDP][Easy] Update `state_dict()` docstring
* **#77850 [FSDP][Easy] Reword device placement warning**
* #77848 [FSDP][Easy] Fix `state_dict_type()` docstring example
* #77847 [FSDP][Easy] Doc fixes

Following up on my comment here: https://github.com/pytorch/pytorch/pull/77321#discussion_r872690381

Feel free to suggest changes as well.

Commit:
[FSDP][Easy] Reword device placement warning

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Reword device placement warning"


Following up on my comment here: https://github.com/pytorch/pytorch/pull/77321#discussion_r872690381

Feel free to suggest changes as well.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -439,14 +439,13 @@ class FullyShardedDataParallel(nn.Module):
+        ``module`` should already be placed on that device, (2) the device
+        should be set using ``torch.cuda.set_device(dev_id)``, or (3)
+        ``dev_id`` should be passed into the ``device_id`` constructor
+        argument. This FSDP instance's compute device is CPU only if module is
```
I think if we check the impl of `get_compute_device` it is never CPU? If the module is input on CPU + device_id is not specified, we return the current CUDA device.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -439,14 +439,13 @@ class FullyShardedDataParallel(nn.Module):
-        that if ``device_id`` is specified, compute device will always match
-        ``device_id``.
+        If the destination CUDA device has ID ``dev_id``, either (1)
+        ``module`` should already be placed on that device, (2) the device
```
I think doing (2) is actually different from doing (1) or (3).

If we do (2) and module is on CPU, sharding/init will still run on CPU. But for (1) or (3), it would run on GPU.

rohan-varma reviewed (commented):
I think we can enhance the doc a little bit

Commit:
Update on "[FSDP][Easy] Reword device placement warning"


Following up on my comment here: https://github.com/pytorch/pytorch/pull/77321#discussion_r872690381

Feel free to suggest changes as well.

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Reword device placement warning"


Following up on my comment here: https://github.com/pytorch/pytorch/pull/77321#discussion_r872690381

Feel free to suggest changes as well.

[ghstack-poisoned]

Commit:
Update on "[FSDP][Easy] Reword device placement warning"


Following up on my comment here: https://github.com/pytorch/pytorch/pull/77321#discussion_r872690381

Feel free to suggest changes as well.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge this on green
